[iOS] Add experimental support for EID/IMEI AutoFill in WKWebView

whsieh opened the pull request:
#### 64d46bbaf793de4e47e68cda1e36119c3832480f
<pre>
[iOS] Add experimental support for EID/IMEI AutoFill in WKWebView
<a href="https://bugs.webkit.org/show_bug.cgi?id=263831">https://bugs.webkit.org/show_bug.cgi?id=263831</a>
rdar://116011445

Reviewed by Richard Robinson and Abrar Rahman Protyasha.

Add support for the `device-eid` and `device-imei` values for the `autocomplete` attribute, by
plumbing these values over to the `UITextContentTypeCellularEID` and `UITextContentTypeCellularIMEI`
platform text content types, respectively.

Test: KeyboardInputTests.DeviceEIDAndIMEIAutoFill

Source/WTF/wtf/PlatformHave.h:

Add a new compile-time flag to guard availability of the CoreTelephony framework.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.mm: Added.
* Source/WebCore/PAL/pal/spi/cocoa/CoreTelephonySPI.h: Added.

Add new SPI headers and soft-linking support for the CoreTelephony framework.

* Source/WebCore/html/Autofill.cpp:
* Source/WebCore/html/Autofill.h:

Add support for the new values.

* Source/WebKit/Platform/Logging.h:

Add a `Telephony` log channel.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h: Added.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm: Added.

Call out to the CoreTelephony framework to query whether or not the current active URL is allowed
for the purposes of ESIM AutoFill.

(WebKit::shouldAllowAutoFillForCellularIdentifiers):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::shouldAllowAutoFillForCellularIdentifiers const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::pageLoadState const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView contentTypeFromFieldName:]):

Move this static helper into `WKContentView` as a regular method, so that we can call into
`WebPageProxy` and return either `nil` or the appropriate text content type, in the case of these
new `device-*` AutoFill types.

(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView textContentTypeForQuickboard]):
(contentTypeFromFieldName): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-iOS.entitlements:

Add an `com.apple.CommCenter.fine-grained` entitlement, so that we can test this new codepath.

* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::allowESIMAutoFillForWebKitDomains):

Add a new API test to exercise the scenarios where CoreTelephony accepts or rejects AutoFill,
depending on the current domain.

Canonical link: <a href="https://commits.webkit.org/269980@main">https://commits.webkit.org/269980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bb6f29bdbe5ac7700fa8f4d5a95112172c11dad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26291 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22249 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3887 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24631 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22713 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24398 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20878 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26882 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21797 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28035 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21008 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22016 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22090 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25812 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23457 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1482 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/19151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30852 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1583 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6770 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5790 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1887 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30816 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1838 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6445 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
```diff
@@ -0,0 +1,114 @@
+        return false;
+    }
+
+    static bool hasPublicCellularPlanEntitlement = [&] {
```
is this actually capturing anything?

whsieh replied:
Good catch! It's not — I'll remove the `&`.

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/CoreTelephonySPI.h:
```diff
@@ -0,0 +1,44 @@
+@end
+#endif
+
+@interface CoreTelephonyClient (Staging_117597722)
```
Optional: put a `FIXME` comment here to remove this declaration eventually?

whsieh replied:
Hm...I think that's implicit with the `Staging_NNNNN` categories (i.e., they should ideally all be removed once the associated radar is in relevant SDKs).

(In any case — still added a `FIXME` here)

aprotyas commented on Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h:
```diff
@@ -0,0 +1,38 @@
+
+namespace WebKit {
+
+#if HAVE(ESIM_AUTOFILL_SYSTEM_SUPPORT)
```
Where is `HAVE_ESIM_AUTOFILL_SYSTEM_SUPPORT` defined?

whsieh replied:
This compile-time flag is not defined in OpenSource.

whsieh committed:
[iOS] Add experimental support for EID/IMEI AutoFill in WKWebView
https://bugs.webkit.org/show_bug.cgi?id=263831
rdar://116011445

Reviewed by Richard Robinson and Abrar Rahman Protyasha.

Add support for the `device-eid` and `device-imei` values for the `autocomplete` attribute, by
plumbing these values over to the `UITextContentTypeCellularEID` and `UITextContentTypeCellularIMEI`
platform text content types, respectively.

Test: KeyboardInputTests.DeviceEIDAndIMEIAutoFill

Source/WTF/wtf/PlatformHave.h:

Add a new compile-time flag to guard availability of the CoreTelephony framework.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/CoreTelephonySoftLink.mm: Added.
* Source/WebCore/PAL/pal/spi/cocoa/CoreTelephonySPI.h: Added.

Add new SPI headers and soft-linking support for the CoreTelephony framework.

* Source/WebCore/html/Autofill.cpp:
* Source/WebCore/html/Autofill.h:

Add support for the new values.

* Source/WebKit/Platform/Logging.h:

Add a `Telephony` log channel.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.h: Added.
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm: Added.

Call out to the CoreTelephony framework to query whether or not the current active URL is allowed
for the purposes of ESIM AutoFill.

(WebKit::shouldAllowAutoFillForCellularIdentifiers):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::shouldAllowAutoFillForCellularIdentifiers const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::pageLoadState const):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView contentTypeFromFieldName:]):

Move this static helper into `WKContentView` as a regular method, so that we can call into
`WebPageProxy` and return either `nil` or the appropriate text content type, in the case of these
new `device-*` AutoFill types.

(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView textContentTypeForQuickboard]):
(contentTypeFromFieldName): Deleted.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI-iOS.entitlements:

Add an `com.apple.CommCenter.fine-grained` entitlement, so that we can test this new codepath.

* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::allowESIMAutoFillForWebKitDomains):

Add a new API test to exercise the scenarios where CoreTelephony accepts or rejects AutoFill,
depending on the current domain.

Canonical link: https://commits.webkit.org/269980@main
